README.md

```diff
@@ -2,302 +2,708 @@
 [![Build Status](https://travis-ci.org/ismasan/parametric.png)](https://travis-ci.org/ismasan/parametric)
 [![Gem Version](https://badge.fury.io/rb/parametric.png)](http://badge.fury.io/rb/parametric)
 
-DSL for declaring allowed parameters with options, regexp pattern and default values.
+Utilities for declaring and manipulating attribute schemas, including custom validations and coercions.
 
-Useful for building self-documeting APIs, search or form objects.
+Useful for building self-documeting APIs, search or form objects. Or possibly as an alternative to Rails' _strong parameters_ (it has no dependencies on Rails and can be used stand-alone).
 
-## Usage
+## Schema
 
-Declare your parameters
+Define a schema
 
 ```ruby
-class OrdersSearch
-  include Parametric::Params
-  param :q, 'Full text search query'
-  param :page, 'Page number', default: 1
-  param :per_page, 'Items per page', default: 30
-  param :status, 'Order status', options: ['checkout', 'pending', 'closed', 'shipped'], multiple: true
+schema = Parametric::Schema.new do
+  field(:title).type(:string).present
+  field(:status).options(["draft", "published"]).default("draft")
+  field(:tags).type(:array)
 end
 ```
 
 Populate and use. Missing keys return defaults, if provided.
 
 ```ruby
-order_search = OrdersSearch.new(page: 2, q: 'foobar')
-order_search.params[:page] # => 2
-order_search.params[:per_page] # => 30
-order_search.params[:q] # => 'foobar'
-order_search.params[:status] # => nil
+form = schema.resolve(title: "A new blog post", tags: ["tech"])
+
+form.output # => {title: "A new blog post", tags: ["tech"], status: "draft"}
+form.errors # => {}
 ```
 
 Undeclared keys are ignored.
 
 ```ruby
-order_search = OrdersSearch.new(page: 2, foo: 'bar')
-order_params.params.has_key?(:foo) # => false
+form = schema.resolve(foobar: "BARFOO", title: "A new blog post", tags: ["tech"])
+
+form.output # => {title: "A new blog post", tags: ["tech"], status: "draft"}
 ```
 
+Validations are run and errors returned
+
+
 ```ruby
-order_search = OrderParams.new(status: 'checkout,closed')
-order_search.params[:status] #=> ['checkout', 'closed']
+form = schema.resolve({})
+form.errors # => {"$.title" => ["is required"]}
```

Comment:
Yeah, I sort of deferred that for later but maybe it'd be good to come up with a decent i118n API before releasing.

Maybe the `message` is just an identifier to look the actual message up in i18n, and the default policy messages come with an in-memory lookup for the english copy, but then it can be configured to load locales from YAML or other sources?